<div class="alert alert-success">
<b> Комментарий от ревьюера 2</b>

Привет еще раз. Спасибо, что доделал работу. Оформление комментариев по работе сохраняется. Только обозначим, что это вторая итерация. 

</div>

<div style="border: 2px solid black; padding: 10px; border-radius: 5px; background-color: #f0f0f0;">
    
Привет, Константин.

Меня зовут Людмила, и я буду проводить ревью твоего проекта. 

Предлагаю перейти на «ты», если не против.

Но если это не удобно — дай знать, и мы перейдем на «вы». Моя главная задача — не находить сделанные тобою ошибки, а передать свой опыт и помочь тебе стать аналитиком данных.

Мои комментарии будут в ячейках «Комментарий от ревьюера».

---    
<div class="alert alert-block alert-danger">
    
Критическое замечание, которое следует исправить.
</div> 
    
<div class="alert alert-block alert-warning">
    
Замечание\совет на будущее.
</div>
    
<div class="alert alert-block alert-success">
    
Когда всё сделано правильно.
</div>
    
---        
    
Если ты вносишь изменения в проект по моим рекомендациям, указывай, пожалуйста, это в своих комментариях. Будет удобно, если обозначишь свои комментарии заметным цветом или поместишь 

<div class="alert alert-info">
    
в такую ячейку
</div>

Давай при работе с проектом мы будем общаться с тобой в диалоге, поэтому не удаляй и не перемещай мои комментарии, это поможет мне сделать повторную  проверку проекта быстрее.
___

# Напишите заголовок проекта здесь

- Автор: Гуреев Константин
- Дата: 07.05.2025

### Цели и задачи проекта

<font color='#777778'>Наш цель помочь коллегам "Секретов темнолесья" изучить развитие игровой индустрии. Наша задача предобработать исторические данные, которые коллеги собрали из открытых источников. В них есть информация о продажах игр, сделанных в разных жанрах и выпущенных на разных платформах, а также пользовательские и экспертные оценки игр. </font>

### Описание данных

<font color='#777778'>Данные `/datasets/new_games.csv` содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:
- `Name` — название игры.
- `Platform` — название платформы.
- `Year of Release` — год выпуска игры.
- `Genre` — жанр игры.
- `NA sales` — продажи в Северной Америке (в миллионах проданных копий).
- `EU sales` — продажи в Европе (в миллионах проданных копий).
- `JP sales` — продажи в Японии (в миллионах проданных копий).
- `Other sales` — продажи в других странах (в миллионах проданных копий).
- `Critic Score` — оценка критиков (от 0 до 100).
- `User Score` — оценка пользователей (от 0 до 10).
- `Rating` — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.</font>

### Содержимое проекта

<font color='#777778'>   

- Загрузка и знакомство с данными.
- Проверка ошибок в данных и их предобработка.
- Фильтрация данных.
- Категоризация данных.</font>

---

<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b> 

Описание проекта составлено хорошо, чувствуется внимательный подход к постановке целей и задач. Здорово, что есть чёткая формулировка цели, описание датасета и основных шагов проекта.
</div>

## 1. Загрузка данных и знакомство с ними

- Загрузите необходимые библиотеки Python и данные датасета `/datasets/new_games.csv`.


In [1]:
# загружаем библиотеку pandas
import pandas as pd

In [2]:
# загружаем данные датасета
df = pd.read_csv('/datasets/new_games.csv')

- Познакомьтесь с данными: выведите первые строки и результат метода `info()`.


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [4]:
df.head()

Name Platform  Year of Release         Genre  NA sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

  EU sales JP sales  Other sales  Critic Score User Score Rating  
0    28.96     3.77         8.45          76.0          8      E  
1     3.58     6.81         0.77           NaN        NaN    NaN  
2    12.76     3.79         3.29          82.0        8.3      E  
3    10.93     3.28         2.95          80.0          8      E  
4     8.89    10.22         1.00           NaN        NaN    NaN

<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b> 

Основная информация о таблице выведена, супер.
</div>

- Сделайте вывод о полученных данных: данные какого объёма вам предоставили, соответствуют ли они описанию, встречаются ли в них пропуски, используются ли верные типы данных.
- Отметьте другие особенности данных, которые вы обнаружили и на которые стоит обратить внимание при предобработке. Например, вы можете проверить названия столбцов: все ли названия отражают содержимое данных и прописаны в удобном для работы виде.

<font color='#777778'>Используйте ячейки типа Markdown для промежуточных выводов и расширенных комментариев к действиям с данными. </font>

<div class="alert alert-info">Данные содержат 11 столбцов и 16956 строк. Из 11 столбцов: 4 содержат данные типа `float64` и 7 `object`.
`Year of Release` стоит перевести в формат `int64` т.к. столбец содержит год(целое число) выпуска игры. `EU sales`, `JP sales`, `User Score`  стоит перевести в формат `float64` т.к. столбцы содержат числа. Названия столбцов необходимо переименовать, приведя названия к snake_case

<div class="alert alert-block alert-warning">
    
<b>Комментарий ревьюера:</b> 

Можно еще указать, что столбцы необходимо переименовать, приведя названия к snake_case.
    
</div>

---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

- Выведите на экран названия всех столбцов датафрейма и проверьте их стиль написания.
- Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания.

In [5]:
df.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

In [6]:
# Переведем названия столбцов в нижний регистр
df.columns = df.columns.str.lower()

In [7]:
# Заменим пробелы в названиях столбцов на подчеркивание
df.columns = df.columns.str.replace(' ', '_')
# выведем результат изменений
df.columns

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')

<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b> 
    
Отлично, столбцы выведены на экран и переименованы.
</div>

### 2.2. Типы данных

- Если встречаются некорректные типы данных, предположите их причины.
- При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу `int64`. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных.

Из 11 столбцов: 4 содержат данные типа `float64` и 7 `object`.
`Year of Release` стоит перевести в формат `int64` т.к. столбец содержит год(целое число) выпуска игры. `EU sales`, `JP sales`, `User Score`  стоит перевести в формат `float64` т.к. столбцы содержат числа.

- В числовых столбцах могут встретиться строковые значения, например `unknown` или другие. Приводите такие столбцы к числовому типу данных, заменив строковые значения на пропуски.

<div class="alert alert-block alert-danger">
    
<b>Комментарий ревьюера:</b> 
    
Перед преобразованием типов данных стоит вывести уникальные значения в столбцах. Так мы сможем увидеть текстовые значения, из-за которых столбец не распознается как числовой, что сделает наше преобразование более обоснованным.
    
</div>

<div class="alert alert-info">
    
Проверим уникальные значения. А также преобразуем столбец `user_score`
</div>

<div class="alert alert-success">
<b> Комментарий от ревьюера 2</b>

Отлично, видны причины неверного распознавания данных.

</div>

In [8]:
# Проверим уникальные значения в столбцах `eu_sales` ,`jp_sales`, `user_score`
unique_eu_sales = df['eu_sales'].unique()
unique_jp_sales = df['jp_sales'].unique()
unique_user_score = df['user_score'].unique()
display(unique_eu_sales, unique_jp_sales, unique_user_score)

array(['28.96', '3.58', '12.76', '10.93', '8.89', '2.26', '9.14', '9.18',
       '6.94', '0.63', '10.95', '7.47', '6.18', '8.03', '4.89', '8.49',
       '9.09', '0.4', '3.75', '9.2', '4.46', '2.71', '3.44', '5.14',
       '5.49', '3.9', '5.35', '3.17', '5.09', '4.24', '5.04', '5.86',
       '3.68', '4.19', '5.73', '3.59', '4.51', '2.55', '4.02', '4.37',
       '6.31', '3.45', '2.81', '2.85', '3.49', '0.01', '3.35', '2.04',
       '3.07', '3.87', '3.0', '4.82', '3.64', '2.15', '3.69', '2.65',
       '2.56', '3.11', '3.14', '1.94', '1.95', '2.47', '2.28', '3.42',
       '3.63', '2.36', '1.71', '1.85', '2.79', '1.24', '6.12', '1.53',
       '3.47', '2.24', '5.01', '2.01', '1.72', '2.07', '6.42', '3.86',
       '0.45', '3.48', '1.89', '5.75', '2.17', '1.37', '2.35', '1.18',
       '2.11', '1.88', '2.83', '2.99', '2.89', '3.27', '2.22', '2.14',
       '1.45', '1.75', '1.04', '1.77', '3.02', '2.75', '2.16', '1.9',
       '2.59', '2.2', '4.3', '0.93', '2.53', '2.52', '1.79', '1.3', '2.6',
   

array(['3.77', '6.81', '3.79', '3.28', '10.22', '4.22', '6.5', '2.93',
       '4.7', '0.28', '1.93', '4.13', '7.2', '3.6', '0.24', '2.53',
       '0.98', '0.41', '3.54', '4.16', '6.04', '4.18', '3.84', '0.06',
       '0.47', '5.38', '5.32', '5.65', '1.87', '0.13', '3.12', '0.36',
       '0.11', '4.35', '0.65', '0.07', '0.08', '0.49', '0.3', '2.66',
       '2.69', '0.48', '0.38', '5.33', '1.91', '3.96', '3.1', '1.1',
       '1.2', '0.14', '2.54', '2.14', '0.81', '2.12', '0.44', '3.15',
       '1.25', '0.04', '0.0', '2.47', '2.23', '1.69', '0.01', '3.0',
       '0.02', '4.39', '1.98', '0.1', '3.81', '0.05', '2.49', '1.58',
       '3.14', '2.73', '0.66', '0.22', '3.63', '1.45', '1.31', '2.43',
       '0.7', '0.35', '1.4', '0.6', '2.26', '1.42', '1.28', '1.39',
       '0.87', '0.17', '0.94', '0.19', '0.21', '1.6', '0.16', '1.03',
       '0.25', '2.06', '1.49', '1.29', '0.09', '2.87', '0.03', '0.78',
       '0.83', '2.33', '2.02', '1.36', '1.81', '1.97', '0.91', '0.99',
       '0.95', '2.0'

array(['8', nan, '8.3', '8.5', '6.6', '8.4', '8.6', '7.7', '6.3', '7.4',
       '8.2', '9', '7.9', '8.1', '8.7', '7.1', '3.4', '5.3', '4.8', '3.2',
       '8.9', '6.4', '7.8', '7.5', '2.6', '7.2', '9.2', '7', '7.3', '4.3',
       '7.6', '5.7', '5', '9.1', '6.5', 'tbd', '8.8', '6.9', '9.4', '6.8',
       '6.1', '6.7', '5.4', '4', '4.9', '4.5', '9.3', '6.2', '4.2', '6',
       '3.7', '4.1', '5.8', '5.6', '5.5', '4.4', '4.6', '5.9', '3.9',
       '3.1', '2.9', '5.2', '3.3', '4.7', '5.1', '3.5', '2.5', '1.9', '3',
       '2.7', '2.2', '2', '9.5', '2.1', '3.6', '2.8', '1.8', '3.8', '0',
       '1.6', '9.6', '2.4', '1.7', '1.1', '0.3', '1.5', '0.7', '1.2',
       '2.3', '0.5', '1.3', '0.2', '0.6', '1.4', '0.9', '1', '9.7'],
      dtype=object)

In [9]:
# `EU sales`, `JP sales` преобразуем во float64 т.к. они не содержат пропуски(ипользуем метод to_numeric т.к. столбци содержат строковые значения)
df['eu_sales'] = pd.to_numeric(df['eu_sales'],errors='coerce')
df['jp_sales'] = pd.to_numeric(df['jp_sales'],errors='coerce')
df['user_score'] = pd.to_numeric(df['user_score'], errors='coerce')

In [10]:
# проверем внесенные изменения
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16681 non-null  float64
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float64
 5   eu_sales         16950 non-null  float64
 6   jp_sales         16952 non-null  float64
 7   other_sales      16956 non-null  float64
 8   critic_score     8242 non-null   float64
 9   user_score       7688 non-null   float64
 10  rating           10085 non-null  object 
dtypes: float64(7), object(4)
memory usage: 1.4+ MB


<div class="alert alert-block alert-warning">
    
<b>Комментарий ревьюера:</b> 
    
Изменение формата данных в столбце с оценками пользователей также стоит перенести сюда, чтобы в следующем пункте верно посчиталось количество пропусков.
    
</div>

### 2.3. Наличие пропусков в данных

- Посчитайте количество пропусков в каждом столбце в абсолютных и относительных значениях.


In [11]:
#посчитаем пропуски по каждому столбцу
df.isna().sum()

name                  2
platform              0
year_of_release     275
genre                 2
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score       8714
user_score         9268
rating             6871
dtype: int64

In [12]:
# опредлелим пропуски в долях
df.isna().sum() / len(df)

name               0.000118
platform           0.000000
year_of_release    0.016218
genre              0.000118
na_sales           0.000000
eu_sales           0.000354
jp_sales           0.000236
other_sales        0.000000
critic_score       0.513918
user_score         0.546591
rating             0.405225
dtype: float64

In [13]:
# в процентах:
df.isna().sum() / len(df) * 100

name                0.011795
platform            0.000000
year_of_release     1.621845
genre               0.011795
na_sales            0.000000
eu_sales            0.035386
jp_sales            0.023590
other_sales         0.000000
critic_score       51.391838
user_score         54.659118
rating             40.522529
dtype: float64

<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b> 
    
Количество и доля пропусков рассчитаны корректно. 
</div>

<div class="alert alert-block alert-warning">
    
<b>Комментарий ревьюера:</b> 
    
Для большей наглядности можно еще отсортировать результаты расчета.
</div>

- Изучите данные с пропущенными значениями. Напишите промежуточный вывод: для каких столбцов характерны пропуски и сколько их. Предположите, почему пропуски могли возникнуть. Укажите, какие действия с этими данными можно сделать и почему.


В столбцах             
- `critic_score`       51% пропусков
- `user_score`         54% пропусков
- `rating`             40% пропусков

Большое число пропусков в рейтигах игр может быть связано с годом выпуска (до 2000). Пропуски в данных столбцах необходимо обработать.

В остальных столбцах пропусков не значительное кол-во (менее 5%). Данные пропуски можно удалить т.к. они не сильно повлияют на анализ.

- Обработайте пропущенные значения. Для каждого случая вы можете выбрать оптимальный, на ваш взгляд, вариант: заменить на определённое значение, оставить как есть или удалить.
- Если вы решите заменить пропуски на значение-индикатор, то убедитесь, что предложенное значение не может быть использовано в данных.
- Если вы нашли пропуски в данных с количеством проданных копий игры в том или ином регионе, их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

<div class="alert alert-info">
столбцы `name`, `genre` можем оставить без изменений т.к. имеют незначительное число пропусков.

в столбце `year_of_release` пропуски можно удалить т.к. они не попадут в срез по этому столбцу.

столбцы `eu_sales` и `jp_sales` заполним средним значением в зависимости от платформы и года выхода игры без изменений т.к. имеют незначительное число пропусков.
</div>

In [14]:
# Сохраним кол-во строк до удаления
df_original_rows = df.shape[0]
# Заполним пропуски в столбцах `eu_sales` и `jp_sales` с помощью apply
# Заполнение пропусков средними значениями по группе (платформа + год)

def fill_missing(group):
    group['eu_sales'] = group['eu_sales'].fillna(group['eu_sales'].mean())
    group['jp_sales'] = group['jp_sales'].fillna(group['jp_sales'].mean())
    return group
# Применяем группировку
df = df.groupby(['platform', 'year_of_release']).apply(fill_missing)
# Кол-во строк после удаления
df_no_null = df.shape[0]
# Проверим результаты
df.isna().sum()

name                  2
platform              0
year_of_release       0
genre                 2
na_sales              0
eu_sales              0
jp_sales              0
other_sales           0
critic_score       8596
user_score         9123
rating             6780
dtype: int64

<div class="alert alert-info">
    После группировки по году пропуски в столбце `year_of_release` (275 пропусков) были автоматически удалены 

<div class="alert alert-block alert-danger">
    
<b>Комментарий ревьюера:</b> 
    
В задании нас просят заменить пропуски в столбцах с продажами jp_sales и eu_sales на среднее значение в зависимости от названия платформы и года выхода игры. Заполнение средним по группам позволяет сохранить данные ближе к их исходной структуре. Это особенно важно для последующего анализа, например, если нужно исследовать динамику продаж по годам или платформам.

</div>

<div class="alert alert-success">
<b> Комментарий от ревьюера 2</b>

Отлично, это заполнение, которое нас просят сделать в задании.

</div>

В столбцах `critic_score`, `user_score`, необходимо заменить пропуски на  определенное значение т.к. они будут использоваться для категоризации.

`rating` оставим без изменений т.к. эти данные не нужны для анализа

<div class="alert alert-block alert-warning">
    
<b>Комментарий ревьюера:</b> 
    
Столбец с рейтингом тоже можно заполнить, так как предполагается, что мы делаем только предобработку, а анализ будет дальше.
    
</div>

In [15]:
# Для user_score
df['user_score'] = df['user_score'].fillna(-1)
df['user_score'] = df['user_score'].astype('float64')
# Для critic_score
df['critic_score'] = df['critic_score'].fillna(-1)
df['critic_score'] = df['critic_score'].astype('int64')

<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b> 
    
Да, эти пропуски можно заполнить значениями-индикаторами.
</div>

In [16]:
# переведем столбец в формат int64
df['year_of_release'] = df['year_of_release'].astype('int64')

<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b> 
    
Да, тут процент пропусков небольшой, поэтому можно просто удалить данные строки.
</div>

In [17]:
# проверим результат
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16681 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16679 non-null  object 
 1   platform         16681 non-null  object 
 2   year_of_release  16681 non-null  int64  
 3   genre            16679 non-null  object 
 4   na_sales         16681 non-null  float64
 5   eu_sales         16681 non-null  float64
 6   jp_sales         16681 non-null  float64
 7   other_sales      16681 non-null  float64
 8   critic_score     16681 non-null  int64  
 9   user_score       16681 non-null  float64
 10  rating           9901 non-null   object 
dtypes: float64(5), int64(2), object(4)
memory usage: 2.0+ MB


<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b> 
    
Молодец, что проверяешь данные после преобразования.

</div>

### 2.4. Явные и неявные дубликаты в данных

- Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
- При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

In [18]:
# проверим уникальные значения платформ
unique_plaform = df['platform'].unique()
display(unique_plaform)

array(['Wii', 'NES', 'GB', 'DS', 'X360', 'PS3', 'PS2', 'SNES', 'GBA',
       'PS4', '3DS', 'N64', 'PS', 'XB', 'PC', '2600', 'PSP', 'XOne',
       'WiiU', 'GC', 'GEN', 'DC', 'PSV', 'SAT', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

In [19]:
# проверим уникальные значения жанра игры
unique_genre = df['genre'].unique()
display(unique_genre)

array(['Sports', 'Platform', 'Racing', 'Role-Playing', 'Puzzle', 'Misc',
       'Shooter', 'Simulation', 'Action', 'Fighting', 'Adventure',
       'Strategy', nan, 'MISC', 'ROLE-PLAYING', 'RACING', 'ACTION',
       'SHOOTER', 'FIGHTING', 'SPORTS', 'PLATFORM', 'ADVENTURE',
       'SIMULATION', 'PUZZLE', 'STRATEGY'], dtype=object)

In [20]:
# проверим уникальные заначения рейтинга
unique_rating = df['rating'].unique()
display(unique_rating)

array(['E', nan, 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP'], dtype=object)

Мы наблюдаем наличие неявных дубликатов в столбце с жанрами игры в связи с написание названий в разных регистрах.

In [21]:
# приведем значения жанров игры к нижнему регистру
df['genre'] = df['genre'].str.lower()
# проверим результат
unique_genre = df['genre'].unique()
display(unique_genre)

array(['sports', 'platform', 'racing', 'role-playing', 'puzzle', 'misc',
       'shooter', 'simulation', 'action', 'fighting', 'adventure',
       'strategy', nan], dtype=object)

<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b> 
    
Хороший подход с приведением к одному регистру, что позволяет избежать неявных дубликатов.
</div>

- После того как нормализуете данные и устраните неявные дубликаты, проверьте наличие явных дубликатов в данных.

In [22]:
# Сохраняем количество строк до удаления дубликатов
initial_row_count = df.shape[0]

# Сортируем датафрем по всем столбцам
df_sorted = df.sort_values(by=list(df.columns))

# Удаляем дубликаты
df_no_duplicates = df_sorted.drop_duplicates()

# Сохраняем количество строк после удаления дубликатов
final_row_count = df_no_duplicates.shape[0]
df_duplicates = initial_row_count - final_row_count
display(f'кол-во строк до удаления дубликатов {initial_row_count}, после удаления {final_row_count} удалено {df_duplicates}')


'кол-во строк до удаления дубликатов 16681, после удаления 16446 удалено 235'

- Напишите промежуточный вывод: укажите количество найденных дубликатов и действия по их обработке.

После обработки неявных дубликатов в столбце `genre`, было выявлено 235 явных дубликатов, после выявления они были удалены. 

- В процессе подготовки данных вы могли что-либо удалять, например строки с пропусками или ошибками, дубликаты и прочее. В этом случае посчитайте количество удалённых строк в абсолютном и относительном значениях.

In [23]:
# Кол-вл строк удаленных при обработке пропусков
df_delet_row = 275
# всего удалено строк 
total_row = df_duplicates + df_delet_row
# Доля удаленных строк
share_row_delet = total_row / df_original_rows
display(f'Удалено {total_row} строк из {df_original_rows}. Доля удаленнных строк {share_row_delet}')

'Удалено 510 строк из 16956. Доля удаленнных строк 0.030077848549186128'

<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b> 
    
Здорово, что ты посчитал количество и долю удаленных строк.
</div>

- После проведения предобработки данных напишите общий промежуточный вывод.

Мы провели преобработку данных были обработаны пропуски и дубликаты в данных. Всего при обработке было удалено 510 строк. Неявные дубликаты в стобце `genre` были обработаны путем приведения строков данных в нижний регистр. Также названия столбцов были приведены к формату snake_case, изменены типы данных в столбцах: 
 - `year_of_release`, `critic_score` на `int64` 
 - `eu_sales`, `jp_sales`, `user_score` на `float64`


<div class="alert alert-block alert-warning">
    
<b>Комментарий ревьюера:</b> 
    
Вывод стоит дополнить, указав не только количество удаленных данных, но и результаты изменения формата, обработки пропусков и неявных дубликатов, а также то, что были скорректированы названия столбцов.
</div>

---

## 3. Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

In [24]:
# созданим срез данных по столбцу `year_of_release` в новом датафрейме df_actual.
df_actual = df[(df['year_of_release'] >= 2000) & (df['year_of_release'] <= 2013)].copy()
# Выводим информацию о новом срезе
df_actual.info()
df_actual.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12980 entries, 0 to 16954
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             12980 non-null  object 
 1   platform         12980 non-null  object 
 2   year_of_release  12980 non-null  int64  
 3   genre            12980 non-null  object 
 4   na_sales         12980 non-null  float64
 5   eu_sales         12980 non-null  float64
 6   jp_sales         12980 non-null  float64
 7   other_sales      12980 non-null  float64
 8   critic_score     12980 non-null  int64  
 9   user_score       12980 non-null  float64
 10  rating           8847 non-null   object 
dtypes: float64(5), int64(2), object(4)
memory usage: 1.2+ MB


name platform  year_of_release     genre  na_sales  \
0             Wii Sports      Wii             2006    sports     41.36   
2         Mario Kart Wii      Wii             2008    racing     15.68   
3      Wii Sports Resort      Wii             2009    sports     15.61   
6  New Super Mario Bros.       DS             2006  platform     11.28   
7               Wii Play      Wii             2006      misc     13.96   

   eu_sales  jp_sales  other_sales  critic_score  user_score rating  
0     28.96      3.77         8.45            76         8.0      E  
2     12.76      3.79         3.29            82         8.3      E  
3     10.93      3.28         2.95            80         8.0      E  
6      9.14      6.50         2.88            89         8.5      E  
7      9.18      2.93         2.84            58         6.6      E

<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b> 
    
Отлично, фильтрация соответствует требованиям задания.
</div>

<div class="alert alert-block alert-warning">
    
<b>Комментарий ревьюера:</b> 
    
Здесь можно дополнительно подсчитать количество игр по годам. Это позволит выявить тенденции в объёме выпускаемых игр: были ли годы с пиками или спадами в количестве выпущенных игр, поможет понять, насколько данные сбалансированы и можно ли делать обобщения для всего периода.
</div>

In [25]:
# Подсчитаем кол-во игр по годам
df_actual.groupby('year_of_release')['name'].count()

year_of_release
2000     357
2001     491
2002     839
2003     789
2004     771
2005     950
2006    1020
2007    1218
2008    1445
2009    1450
2010    1279
2011    1149
2012     670
2013     552
Name: name, dtype: int64

<div class="alert alert-info">
    Наиболее приятный год для игроков был 2009 было выпущено 1450 игр

---

## 4. Категоризация данных
    
Проведите категоризацию данных:
- Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

In [26]:
# Созданим новый столец с категориями по оценкам пользователей
df_actual['category_user_score'] = pd.cut(df['user_score'], bins=[0,3,8,10], labels=['низкая оценка','средняя оценка','высокая оценка'], right=False)

In [27]:
# Проверим результат
df_actual.head()

name platform  year_of_release     genre  na_sales  \
0             Wii Sports      Wii             2006    sports     41.36   
2         Mario Kart Wii      Wii             2008    racing     15.68   
3      Wii Sports Resort      Wii             2009    sports     15.61   
6  New Super Mario Bros.       DS             2006  platform     11.28   
7               Wii Play      Wii             2006      misc     13.96   

   eu_sales  jp_sales  other_sales  critic_score  user_score rating  \
0     28.96      3.77         8.45            76         8.0      E   
2     12.76      3.79         3.29            82         8.3      E   
3     10.93      3.28         2.95            80         8.0      E   
6      9.14      6.50         2.88            89         8.5      E   
7      9.18      2.93         2.84            58         6.6      E   

  category_user_score  
0      высокая оценка  
2      высокая оценка  
3      высокая оценка  
6      высокая оценка  
7      средняя оценка

<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b> 
    
Отличная работа! Ты использовал метод pd.cut(), что позволило упростить код. Это демонстрирует твоё хорошее знание библиотеки Pandas и умение применять встроенные функции для оптимизации работы с данными.
    
</div>

<div class="alert alert-block alert-warning">
    
<b>Комментарий ревьюера:</b> 
    
Также стоит добавить категорию "без оценки", так как в данных столбцах встречаются пропуски, заполненные значением -1.
</div>

- Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

In [28]:
# Созданим новый столец с категориями по оценкам критиков
df_actual['category_critic_score'] = pd.cut(df['critic_score'], bins=[0,30,80,100], labels=['низкая оценка','средняя оценка','высокая оценка'], right=False)

In [29]:
# Проверим результат
df_actual.head()

name platform  year_of_release     genre  na_sales  \
0             Wii Sports      Wii             2006    sports     41.36   
2         Mario Kart Wii      Wii             2008    racing     15.68   
3      Wii Sports Resort      Wii             2009    sports     15.61   
6  New Super Mario Bros.       DS             2006  platform     11.28   
7               Wii Play      Wii             2006      misc     13.96   

   eu_sales  jp_sales  other_sales  critic_score  user_score rating  \
0     28.96      3.77         8.45            76         8.0      E   
2     12.76      3.79         3.29            82         8.3      E   
3     10.93      3.28         2.95            80         8.0      E   
6      9.14      6.50         2.88            89         8.5      E   
7      9.18      2.93         2.84            58         6.6      E   

  category_user_score category_critic_score  
0      высокая оценка        средняя оценка  
2      высокая оценка        высокая оценка  
3      высокая оценка        высокая оценка  
6      высокая оценка        высокая оценка  
7      средняя оценка        средняя оценка

- После категоризации данных проверьте результат: сгруппируйте данные по выделенным категориям и посчитайте количество игр в каждой категории.

In [30]:
# Посчитаем кол-во игр по категориям оценки пользователей
df_actual.groupby('category_user_score')['name'].count()

category_user_score
низкая оценка      117
средняя оценка    4148
высокая оценка    2307
Name: name, dtype: int64

In [31]:
# Посчитаем кол-во игр по категориям оценки критиков 
df_actual.groupby('category_critic_score')['name'].count()

category_critic_score
низкая оценка       55
средняя оценка    5500
высокая оценка    1712
Name: name, dtype: int64

<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b> 
    
Отлично, у нас есть игры во всех категориях.
</div>

In [32]:
# Проведем группировку по обеим категориям
df_actual.groupby(['category_user_score','category_critic_score'])['name'].count()

category_user_score  category_critic_score
низкая оценка        низкая оценка              17
                     средняя оценка             78
                     высокая оценка              1
средняя оценка       низкая оценка              30
                     средняя оценка           3210
                     высокая оценка            649
высокая оценка       низкая оценка               1
                     средняя оценка           1175
                     высокая оценка           1028
Name: name, dtype: int64

- Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [33]:
# Проведем группировку данных по платформам
group_platform = df_actual.groupby('platform').agg({'name':'count'})

In [34]:
# проведем сортировку результата и выведем топ-7 платформ по кол-во игр
top_platform = group_platform.sort_values('name', ascending=False)
top_platform.head(7)

name
platform      
PS2       2154
DS        2146
Wii       1294
PSP       1199
X360      1138
PS3       1107
GBA        826

<div class="alert alert-block alert-success">
    
<b>Комментарий ревьюера:</b> 
    
Топ-7 платформ определены верно.
</div>

---

## 5. Итоговый вывод

В конце напишите основной вывод и отразите, какую работу проделали. Не забудьте указать описание среза данных и новых полей, которые добавили в исходный датасет.

Мы провели преобработку данных были обработаны пропуски и дубликаты в данных. Всего при обработке было удалено 510 строк. Неявные дубликаты в стобце `genre` были обработаны путем приведения строков данных в нижний регистр.

Для анализа выпущенных игр за первую половину 21 века мы создали срез данных по столбцу `year_of_release` c 2000 по 2013 года и сохранили в новый датасет `df_actual`.
- `name` — название игры.
- `platform` — название платформы.
- `year_of_release` — год выпуска игры.
- `genre` — жанр игры.
- `na_sales` — продажи в Северной Америке (в миллионах проданных копий).
- `eu_sales` — продажи в Европе (в миллионах проданных копий).
- `jp_sales` — продажи в Японии (в миллионах проданных копий).
- `other_sales` — продажи в других странах (в миллионах проданных копий).
- `critic_score` — оценка критиков (от 0 до 100).
- `user_score` — оценка пользователей (от 0 до 10).
- `rating` — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.
- `category_user_score` - категория по оценкам пользователей
- `category_critic_score` - категория по оценкам критиков

В новом датасете мы добавили новые столбцы с категориями `category_user_score` и `category_critic_score` по оценкам пользователей и критиком соответственно. Далее мы провели группировку данных по данным столбцам и подсчитали для каждой группы ('низкая оценка','средняя оценка','высокая оценка') кол-во выпущенных игр.

category_user_score  кол-во игр                    

низкая оценка      -      55                        

средняя оценка      -    5500                       

высокая оценка      -    1712                       
                       
category_critic_score    кол-во игр

низкая оценка        -       117

средняя оценка       -       4148

высокая оценка       -       2307

А также вывели Топ-7 платформ по кол-ву выпущенных игр за период 2000-2013г.

платформа     кол-во игр
 
 PS2	        - 2154
  
  DS	        - 2146
  
  Wii	        - 1294
  
  PSP	        - 1199
  
  X360	        - 1138
  
  PS3	        - 1107
  
  GBA	        - 826

<div class="alert alert-block alert-warning">
    
<b>Комментарий ревьюера:</b>  
    
Здесь можно еще добавить обновлённое описание полей итогового датафрейма. Это хороший способ показать, что данные были полностью подготовлены и какие изменения внесены в структуру. 
</div>

<div style="border:solid Chocolate 2px; padding: 40px">

# <b>Итоговый комментарий ревьюера:</b>
    
<div class="alert alert-success">

Благодарю тебя за выполнение проекта. Видно, что ты постарался, когда работал над ним. Однако есть несколько моментов, которые необходимо дополнить:
    
* вывести уникальные значения в столбцах перед заменой;
* дополнить обработку пропусков.

Жду проект после доработки. Уверена, ты справишься. 💪
</div>

<div class="alert alert-success">
<b> Комментарий от ревьюера 2</b>

Спасибо за внимательное отношение к комментариям. Проект принят, и желаю успехов в дальнейшем обучении на курсе!
    
</div>